spark3/src/main/java/org/apache/iceberg/spark/FileRewriteCoordinator.java

```diff
@@ -0,0 +1,169 @@
+  public static FileRewriteCoordinator get() {
+    return INSTANCE;
+  }
+
+  public void stageRewrite(Table table, String fileSetID, Set<DataFile> newDataFiles) {
+    Preconditions.checkArgument(newDataFiles != null && newDataFiles.size() > 0, "Cannot stage null or empty files");
+    LOG.info("Staging results of rewriting file set {} for table {}", fileSetID, table);
+    Pair<String, String> id = toID(table, fileSetID);
+    resultMap.put(id, newDataFiles);
+  }
+
+  public void commitRewrite(Table table, String fileSetID) {
+    commitRewrite(table, ImmutableSet.of(fileSetID));
+  }
+
+  public void commitRewrite(Table table, Set<String> fileSetIDs) {
+    LOG.info("Committing rewrite of file sets {} for table {}", fileSetIDs, table);
+
+    Set<DataFile> rewrittenDataFiles = fetchRewrittenDataFiles(table, fileSetIDs);
+    Set<DataFile> newDataFiles = fetchNewDataFiles(table, fileSetIDs);
+
+    String tableUUID = tableUUID(table);
+    Lock tableLock = lockMap.computeIfAbsent(tableUUID, uuid -> new ReentrantLock());
+
+    try {
+      LOG.info("Locking table {} to commit rewrite of file sets {}", table, fileSetIDs);
+      tableLock.lock();
+
+      table.newRewrite()
+          .rewriteFiles(rewrittenDataFiles, newDataFiles)
+          .commit();
+
+      LOG.info("Successfully committed rewrite of file sets {} for table {}", fileSetIDs, table);
+
+    } finally {
+      LOG.info("Unlocking table {}", table);
+      tableLock.unlock();
+    }
+  }
+
+  private Set<DataFile> fetchRewrittenDataFiles(Table table, Set<String> fileSetIDs) {
+    FileScanTaskSetManager taskSetManager = FileScanTaskSetManager.get();
+
+    Set<DataFile> rewrittenDataFiles = Sets.newHashSet();
+
+    for (String fileSetID : fileSetIDs) {
+      List<FileScanTask> tasks = taskSetManager.fetchTasks(table, fileSetID);
+      ValidationException.check(tasks != null,
+          "Task set manager has no tasks for table %s with id %s",
+          table, fileSetID);
+
+      for (FileScanTask task : tasks) {
+        DataFile dataFile = task.file();
+        rewrittenDataFiles.add(dataFile);
+      }
+    }
+
+    return rewrittenDataFiles;
+  }
+
+  private Set<DataFile> fetchNewDataFiles(Table table, Set<String> fileSetIDs) {
+    List<Set<DataFile>> results = Lists.newArrayList();
+
+    for (String fileSetID : fileSetIDs) {
+      Pair<String, String> id = toID(table, fileSetID);
+      Set<DataFile> result = resultMap.get(id);
+      ValidationException.check(result != null,
+          "No results for compacting file set %s in table %s",
+          fileSetID, table);
+
+      results.add(result);
+    }
+
+    Set<DataFile> newDataFiles = results.get(0);
+    for (int index = 1; index < results.size(); index++) {
+      newDataFiles = Sets.union(newDataFiles, results.get(index));
+    }
+
+    return newDataFiles;
+  }
+
+  public void abortRewrite(Table table, String fileSetID) {
+    LOG.info("Aborting compaction of file set {} for table {}", fileSetID, table);
```

Comment:
Probably should be at the compaction code level?

Reply:
I'll be OK if we want to move all logging to the action. Do you think it will be better? We no longer have locking so I think that's going to be fine.

Reply:
Removed the logging, let's have it at the action level.